### DISCUSSION TITLE
useTreeData -> Selections

### DISCUSSION THREAD

**Discussion by lienas** [adobe/react-spectrum#3595] [Q&A] [answered]
Is there an example on how to use `setSelectedKeys`-function  and `selectedKeys`-prop ??

**Comment by reidbarber** [comment]
The ComboBox docs use it in the [Fully controlled example](https://react-spectrum.adobe.com/react-spectrum/ComboBox.html#fully-controlled-combobox).

**Reply by lienas** [reply]
Hi @reidbarber  Unfortunately that is not what I am looking for.

I think `useTreeData` maintains the state itself. The linked example stores one selected item with `useState` hook.
I guess `setSelectedKeys(keys: Set(<Key>): void`  populates  `selectedKeys | Set<Key>` and therefore `useState` hook is not necessary - is it ? 
What I not know is how to deal with `Set<Key>`

I want to build a tree-view with multiselect !

**Reply by reidbarber** [reply]
I think the only other examples we have are on the [useTreeData](https://react-spectrum.adobe.com/react-stately/useTreeData.html) and [Selection](https://react-spectrum.adobe.com/react-stately/selection.html) pages. You could combine those with the example components on [useListbox](https://react-spectrum.adobe.com/react-aria/useListBox.html) to get an example.

I think we will have more examples once we implement TreeView.

Are there any specific issues you're having?

**Reply by lienas** [reply]
In the meantime a found the following solution, but  when I call `setSelectedKeys` in my `Categories`-component `selectedKeys` of my TreeData is always 1 Click behind although my `onSelectionChangeHandler` receives the correct keys.

Here is my Code:

```
import React from 'react';

import TreeView from "../components-proto/tree-view";
import {Item} from "./search-box";
import {useTreeData} from "react-stately";
import {convertedLiferayCategories} from "../migration/categories.helpers";

const Categories = () => {

    let treeData = useTreeData({
        initialItems: convertedLiferayCategories(),
        getKey: item => item.key,
        getChildren: (item) => item.nodes || []
    })

    const onSelectionChangeHandler = (keys: any) => {
        console.log("Keys ::", keys);
        treeData.setSelectedKeys(keys);
        console.log("treedata -> ", treeData.selectedKeys);
    }

    return (
        <TreeView
            items={treeData.items}
            selectionMode={"multiple"}
            onSelectionChange={onSelectionChangeHandler}
        >
            {treeData.items.map(cat => (
                <Item key={cat.key}>{cat.value.label}</Item>)
            )}
        </TreeView>
    );
};

export default Categories;
```
```
import React from 'react';
import {TreeProps, useTreeState} from 'react-stately';
import {useMenu} from 'react-aria';
import {Button, Checkbox, List, ListItem} from "@chakra-ui/react";

const TreeView = (props: TreeProps<any>) => {
    // Create menu state based on the incoming props
    let state = useTreeState(props);

    // Get props for the menu element
    let ref = React.useRef<HTMLUListElement>(null);
    let {menuProps} = useMenu(props, state, ref);
    const {selectionManager} = state;
    
    const onSelectHandler = (val: any) => {
        selectionManager.toggleSelection(val.key);
    }

    const onSelectAll = () => {
        selectionManager.selectAll();
    }

    return (
        <List
            {...menuProps}
            ref={ref}
        >
            <Button size={"sm"} my={5} onClick={onSelectAll}>all</Button>
            {[...state.collection].map((item) => {
                return (
                    <ListItem key={item.key} alignItems={"center"}>
                        <Checkbox
                            onChange={() => onSelectHandler(item)}
                            value={item.key}
                            isChecked={selectionManager.isSelected(item.key)}
                        >
                            {item.rendered}
                        </Checkbox>
                    </ListItem>
                )
            })}
        </List>
    );
};

export default TreeView;
```
And this is an example output of console.log from the `categories`- component.

<img width="478" alt="Bildschirmfoto 2022-10-08 um 14 13 20" src="https://user-images.githubusercontent.com/8291992/194706998-08be5924-e6dc-4482-bfc9-5336320b2524.png">

**Reply by snowystinger** [reply]
Some quick notes. You shouldn't need to call `treeData.setSelectedKeys(keys);` outside the TreeView. You're already setting selection inside. Also, the onSelectAll should probably be in a check for `selectionManager.isSelectAll`
Would you mind putting together a codesandbox? It'll make debugging a lot easier for us.

**Comment by lienas** [comment, marked as answer]
Hi Robert,

of course:
here weg go: https://codesandbox.io/s/strange-albattani-i1k05k

Thomas


Von: Robert Snow ***@***.***>
Datum: Samstag, 8. Oktober 2022 um 20:42
An: adobe/react-spectrum ***@***.***>
Cc: Thomas Lucas ***@***.***>, Author ***@***.***>
Betreff: Re: [adobe/react-spectrum] useTreeData -> Selections (Discussion #3595)

Some quick notes. You shouldn't need to call treeData.setSelectedKeys(keys); outside the TreeView. You're already setting selection inside. Also, the onSelectAll should probably be in a check for selectionManager.isSelectAll
Would you mind putting together a codesandbox? It'll make debugging a lot easier for us.

**Reply by snowystinger** [reply]
Here, I fixed the codesandbox so it doesn't break. https://codesandbox.io/s/inspiring-margulis-h47mhy?file=/src/tree-view.tsx:688-703

As for why TreeData appears to be one click behind
Because setSelectedKeys, calls a setState in order to updated selectedKeys https://github.com/adobe/react-spectrum/blob/main/packages/%40react-stately/data/src/useTreeData.ts#L134 it won't immediately update
See https://beta.reactjs.org/apis/react/useState#updating-state-based-on-the-previous-state for more information
I've updated the codesandbox I shared earlier to also print out the selectedKeys during render which is called as a result of the setState and it shows the correct value, it's not behind.

Let us know if you have any other questions.

**Reply by lienas** [reply]
@snowystinger Thanks a lot for your fix. So in general my code was working  :-)
There is one more question about the children prop which is passed as an `<Item>` component. Why do I need to pass this child-prop as I also pass the items-prop.  Sorry, but I do not know this pattern!

In my code this looks like the following:
 ```
<TreeView
      items={treeData.items}  <-----
      selectionMode={"multiple"}
      onSelectionChange={onSelectionChangeHandler}
    >
      {treeData.items.map((cat) => (
        <Item key={cat.key}>{cat.value.label}</Item> <------
      ))}
    </TreeView>
```

**Reply by snowystinger** [reply]
Ah, you're mixing two styles. You'll want to read this page https://react-spectrum.adobe.com/react-stately/collections.html#dynamic-collections

**Reply by lienas** [reply]
I made some progress and refactored my code for dynamic data.
Now I added some child-nodes in my data.
```
const CategoriesData = [
  {
    key: "25825",
    label: "Category 1",
    nodes: [
      {
        key: "12345",
        label: "Subcategory 1",
        nodes: []
      }
    ]
  },
...
];
```

In the useTreeData-hook I have the following config:
```
 let treeData = useTreeData({
    initialItems: CategoriesData,
    getKey: (item) => item.key,
    getChildren: (item) => item.nodes || []
  });
```

in the Treeview Component I have:
` let state = useTreeState(props);`

But in the collection of the state the` hasChildNodes` prop is false and the `childNodes`: object is empty.
I have updated my sandbox:  https://codesandbox.io/s/strange-albattani-i1k05k

**Reply by snowystinger** [reply]
I think you're just missing `hasChildItems={cat.nodes}` on your `Item` https://github.com/adobe/react-spectrum/blob/main/packages/%40react-stately/collections/src/Item.ts#L63

**Reply by lienas** [reply]
And what is then the best practice to get the childNotes ?
1. item.childNodes - prop 
    - > I tried to iterate over this prop with `for (let childNode  of item.childNodes)`, but that did not work !
2. item.value.children - prop
    - >  using `item.value.children` works , but is that the recommend way ?

**Reply by snowystinger** [reply]
I think you want to look at this example https://github.com/adobe/react-spectrum/blob/main/packages/%40react-stately/tree/stories/useTreeState.stories.tsx

**Reply by lienas** [reply]
I changed my code to the following:
 ```
<Item key={cat.key}
       hasChildItems={cat.value.nodes.length > 0}
       childItems={cat.nodes}>
```
and in my tree-view I call
```
{[...state.collection].map((item) => {
        if (item.hasChildNodes) {
          console.log("childNodes - >", Array.from(item.childNodes));
        }
        return (...)
```
But I get an empty Array !
I have updated my sandbox!

**Reply by snowystinger** [reply]
@lienas were you able to resolve this? Apologies, I lost track of this discussion, but I know you've been making progress from some other discussions.

**Reply by lienas** [reply]
yes- I could successfully implement selection for my tree-component based on the given story.
